### PR TITLE
Updated account-data-matching,  bump-seed-canonicalization

### PR DESCRIPTION
**Pull request by Epistetechnician**
# Summary of Changes: Bump Seed Canonicalization

## Frontmatter
- **Added** `keywords` and `tags` fields in the new version.
- **Removed** `objectives` field in the new version.

## Summary Section
- **Updated** the summary to be more concise and focused on the key points.

## Lesson Section
- **Expanded** explanation of bump seeds and canonical bumps.
- **Added** detailed examples of insecure and secure PDA derivation.
- **Included** code examples for both insecure and secure implementations.
- **Added** explanation of using Anchor's `seeds` and `bump` constraints.

## Code Examples
- **Added** code examples for:
  - Insecure PDA derivation using `create_program_address`.
  - Secure PDA derivation using `find_program_address`.
  - Using Anchor's `seeds` and `bump` constraints.

## Test Section
- **Added** a detailed test example to show how the secure implementation prevents multiple claims.

## Challenge Section
- **Added** a challenge section encouraging users to audit their own or others' programs for proper PDA derivation.

## Callout Section
- **Added** a callout section prompting users to push their code to GitHub and provide feedback.

### CODE REVIEW

File: content/courses/program-security/account-data-matching.md [hunk 1]
```diff
@@ -427,107 +325,90 @@ pub struct SecureWithdraw<'info> {
...
+  // Check that the withdrawal succeeded
+  const balance = await connection.getTokenAccountBalance(tokenPDA);
+  expect(balance.value.uiAmount).to.eq(0);
+});
 ```
 
-Note that Anchor specifies in the logs the account that causes the error
-(`AnchorError caused by account: vault`).
+Run `anchor test` again. You should see that the unauthorized withdrawal now
+fails, while the authorized one succeeds.
```
Comment (mikemaccana) [inline content/courses/program-security/account-data-matching.md:387]: I lived how the old version showed the test output. It makes the objective clearer for readers. Please re-add.

File: content/courses/program-security/arbitrary-cpi.md [hunk 2]
```diff
@@ -13,20 +13,19 @@ description: "How to safely invoke Solana programs from other Solana programs."
...
-  into the instruction. Your program should check for incorrect or unexpected
-  programs.
-- Perform program checks in native programs by simply comparing the public key
-  of the passed-in program to the progam you expected.
-- If a program is written in Anchor, then it may have a publicly available CPI
+  instruction as an account. Any target program could be passed in, so your
+  program should check for incorrect or unexpected programs.
+- Perform program checks in native programs by comparing the public key of the
+  passed-in program to the program you expected.
+- If a program is written in Anchor, it may have a publicly available CPI
```
Comment (mikemaccana) [inline content/courses/program-security/arbitrary-cpi.md:20]: What is a 'publicly available CPI'? Does it mean 'publicly available IDL' maybe? Or am I missing something?

File: content/courses/program-security/arbitrary-cpi.md [hunk 3]
```diff
@@ -80,59 +79,60 @@ pub mod arbitrary_cpi_insecure {
...
-### Use an Anchor CPI module
+### Use an Anchor CPI Module
 
 A simpler way to manage program checks is to use Anchor CPI modules. We learned
-in a
-[previous lesson](https://github.com/Unboxed-Software/solana-course/blob/main/content/anchor-cpi)
-that Anchor can automatically generate CPI modules to make CPIs into the program
-simpler. These modules also enhance security by verifying the public key of the
-program that’s passed into one of its public instructions.
+in a [previous lesson](../anchor-cpi/README.md) that Anchor can automatically
```
Comment (mikemaccana) [inline content/courses/program-security/arbitrary-cpi.md:131]: Use metadata.yaml to find the actual URL of the previous lesson and a make a link to it per CONTRIBUTING.

File: content/courses/program-security/arbitrary-cpi.md [hunk 4]
```diff
@@ -218,7 +221,7 @@ mints, distribution, and transfers, and a separate metadata program is used to
 assign metadata to tokens. So the vulnerability we go through here could also be
 applied to real tokens.
 
-#### 1. Setup
+### 1. Setup
 
 We'll start with the `starter` branch of
 [this repository](https://github.com/Unboxed-Software/solana-arbitrary-cpi/tree/starter).
```
Comment (mikemaccana) [inline content/courses/program-security/arbitrary-cpi.md:227]: Moved to https://github.com/solana-developers/arbitrary-cpi/

File: content/courses/program-security/arbitrary-cpi.md [hunk 5]
```diff
@@ -397,22 +400,22 @@ pub fn create_character_secure(ctx: Context<CreateCharacterSecure>) -> Result<()
...
-#### 4. Test `create_character_secure`
+### 4. Test `create_character_secure`
 
 Now that we have a secure way of initializing a new character, let's create a
 new test. This test just needs to attempt to initialize the attacker's character
 and expect an error to be thrown.
 
 ```typescript
-it("Secure character creation doesn't allow fake program", async () => {
+it("Secure character creation doesn't allow unauthorized program", async () => {
```
Comment (mikemaccana) [inline content/courses/program-security/arbitrary-cpi.md:410]: Much better. 👍

File: content/courses/program-security/bump-seed-canonicalization.md [hunk 6]
```diff
@@ -1,77 +1,48 @@
...
-nature of allowing multiple bumps.
-
-In the example below, the `set_value` instruction uses a `bump` that was passed
-in as instruction data to derive a PDA. The instruction then derives the PDA
-using `create_program_address` function and checks that the `address` matches
-the public key of the `data` account.
-
-```rust
+Bump seeds are numbers between 0 and 255 used to ensure that an address derived
+using `create_program_address` is a valid PDA. The canonical bump is the highest
```
Comment (mikemaccana) [inline content/courses/program-security/bump-seed-canonicalization.md:33]:
```suggestion
using `create_program_address` is a valid PDA. The **canonical bump** is the highest
```

File: content/courses/program-security/bump-seed-canonicalization.md [hunk 7]
```diff
@@ -1,77 +1,48 @@
...
+valid bump value that produces a valid PDA. For security and consistency, always
+use the canonical bump when deriving PDAs.
+
+### Insecure PDA derivation
+
+The `create_program_address` function produces a valid PDA about 50% of the
+time. Multiple valid bumps can exist for a given set of input seeds, potentially
+causing security issues.
```
Comment (mikemaccana) [inline content/courses/program-security/bump-seed-canonicalization.md:41]: I want some history here. I personally have never heard of `create_program_address`. I probably got into PDAs after everyone stopped using it.

Find out what the history is here, and tell people. Eg:

> You may find older programs using a function called `create_program_address()` to create PDA addresses...

but not these exact words - whatever the actual truth is. 

Also please tell readers:
 - that `create_program_address` is deprecated (if it is deprecated)
 - why `create_program_address` isn't deprecated (if it is not deprecated)

if it's not deprecated and there's no good reason why, file a  big and list it here.

File: content/courses/program-security/bump-seed-canonicalization.md [hunk 8]
```diff
@@ -151,17 +121,17 @@ pub fn set_value_secure(
...
 using the `seeds` and `bump` constraints. These can even be combined with the
 `init` constraint to initialize the account at the intended address. To protect
-the program from the vulnerability we’ve been discussing throughout this lesson,
+the program from the vulnerability we've been discussing throughout this lesson,
 Anchor does not even allow you to initialize an account at a PDA using anything
 but the canonical bump. Instead, it uses `find_program_address` to derive the
 PDA and subsequently performs the initialization.
 
-```rust
+```
```
Comment (mikemaccana) [inline content/courses/program-security/bump-seed-canonicalization.md:134]: Why are you removing syntax highlighting here? Please restore it unless there's a very good reason.

File: content/courses/program-security/bump-seed-canonicalization.md [hunk 9]
```diff
@@ -330,7 +300,7 @@ represent an attacker. It then loops through all possible bumps and calls
 the attacker has been able to claim rewards multiple times and has earned more
 than the 10 tokens allotted per user.
 
-```typescript
+```
```
Comment (mikemaccana) [inline content/courses/program-security/bump-seed-canonicalization.md:303]: Please restore all syntax highlighting.

File: docs/programs/examples.md [hunk 10]
```diff
@@ -37,8 +37,8 @@ with README files that explain you how to run the different examples. Most
 examples are self-contained and are available in native Rust (ie, with no
 framework), [Anchor](https://www.anchor-lang.com/docs/installation),
 [Seahorse](https://seahorse-lang.org/) and it also contains a list of examples
-that we would love to
-[see as contributions](https://github.com/solana-developers/program-examples?tab=readme-ov-file#examples-wed-love-to-see).  
+that we would love to [see as contributions].
```
Comment (mikemaccana) [inline docs/programs/examples.md:40]: why did you remove this link?

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Some good stuff but also a few things that seem odd - please see the comments in this PR.

File: content/courses/program-security/account-data-matching.md [hunk 11]
```diff
@@ -2,59 +2,69 @@
...
+Without proper data validation, your program becomes vulnerable to various
+attacks:
 
-#### Missing data validation check
+1. Unauthorized access: malicious users could pass in unexpected accounts,  
+   gaining access to functionality they shouldn't have
+2. State manipulation: attackers might alter the program's state in unintended  
+   ways, compromising its integrity
+3. Fund theft: in programs dealing with tokens or SOL, inadequate checks could  
+   lead to unauthorized withdrawals
```
[the remark this answers is not in the document]
Comment (Epistetechnician) [reply content/courses/program-security/account-data-matching.md]: Understood

File: content/courses/program-security/account-data-matching.md [hunk 12]
```diff
@@ -76,214 +86,134 @@ pub struct AdminConfig {
...
+of a withdraw instruction to highlight the difference.
 
-```rust
-use anchor_lang::prelude::*;
-use anchor_spl::token::{self, Mint, Token, TokenAccount};
+### 1. Setup
 
-declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
+Clone the starter code from the `starter` branch of
+[this repository](https://github.com/Unboxed-Software/solana-account-data-matching).
```
[the remark this answers is not in the document]
Comment (Epistetechnician) [reply content/courses/program-security/account-data-matching.md]: Got it

File: content/courses/program-security/arbitrary-cpi.md (hunk 2 above)
Comment (Epistetechnician) [reply content/courses/program-security/arbitrary-cpi.md:20]: It is publicly available cross program invocations no?

File: content/courses/program-security/arbitrary-cpi.md (hunk 3 above)
Comment (Epistetechnician) [reply content/courses/program-security/arbitrary-cpi.md]: Got it

File: content/courses/program-security/bump-seed-canonicalization.md (hunk 8 above)
Comment (Epistetechnician) [reply content/courses/program-security/bump-seed-canonicalization.md]: Apologies had forgot to add them back in

File: content/courses/program-security/account-data-matching.md [hunk 13]
```diff
@@ -304,114 +234,82 @@ pub struct InsecureWithdraw<'info> {
...
-    const balance = await connection.getTokenAccountBalance(tokenPDA)
-    expect(balance.value.uiAmount).to.eq(0)
-  })
-})
+test("Insecure withdraw allows unauthorized access", async () => {
+  // Setup: Initialize vault and mint some tokens to it
+  // ... (initialization code here)
+
+  // Attempt unauthorized withdrawal
+  // Note: unauthorizedWallet represents an unauthorized wallet trying to withdraw funds
```
Comment (mikemaccana) [inline content/courses/program-security/account-data-matching.md:256]:
```suggestion
  // Attempt unauthorized withdrawal
```

Honestly the code explains this.

**Comment by mikemaccana** [comment]
@Epistetechnician you're really close to winning multiple prizes here but you have a bunch of references to unboxed repos still in this code:

https://github.com/Unboxed-Software/solana-account-data-matching
https://github.com/Unboxed-Software/solana-arbitrary-cpi

Also https://github.com/Unboxed-Software/solana-course/blob/main/content/anchor-cpi - make this a link to https://solana.com/developers/courses/onchain-development/anchor-cpi in whatever style the links section in CONTRIBUTING.md uses.

**Review by mikemaccana — APPROVED** [review]

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Some minor fixes (see above) still to go then this can go in.

**Comment by mikemaccana** [comment]
> @Epistetechnician you have a bunch of references to unboxed repos still in this code:
> If you fix these I'll mark you as the winner for account-data-matching and bump-seed-canonicalization lessons.

You didn't update the old unboxed repos as requested (and mentioned in the Superteam Bounty), just made links to a new repo that didn't exist. As the bounty has ended, and another user did these changes we will give the prize to the user.
